test/spv-node-sync-test.js

```diff
@@ -0,0 +1,233 @@
+  });
+
+  it('should mine a block', async () => {
+    const block = await miner.mineBlock();
+    assert(block);
+    await chain.add(block);
+
+    // Check SPV & Full nodes are in sync
+    await event(spvnode, 'block');
+    assert.deepStrictEqual(node.chain.tip, spvnode.chain.tip);
+  });
+
+  it('should mine competing chains', async function () {
+    for (let i = 0; i < 10; i++) {
+      const block1 = await mineBlock(tip1);
+      const block2 = await mineBlock(tip2);
+
+      await chain.add(block1);
+      await chain.add(block2);
+
+      assert.bufferEqual(chain.tip.hash, block1.hash());
+
+      tip1 = await chain.getEntry(block1.hash());
+      tip2 = await chain.getEntry(block2.hash());
+
+      assert(tip1);
+      assert(tip2);
+
+      assert(!await chain.isMainChain(tip2));
+
+      await new Promise(setImmediate);
+
+      // Check SPV & Full nodes are in sync after every block
+      await event(spvnode, 'block');
+      assert.deepStrictEqual(node.chain.tip, spvnode.chain.tip);
+    }
+  });
+
+  it('should send a tx from chain 1 to SPV node', async () => {
+    await wallet.send({
+      outputs: [{
+        value: 1012345678,
+        address: spvaddr
+      }]
+    });
+
+    await event(spvwallet, 'balance');
+    const balance = await spvwallet.getBalance();
+    assert.strictEqual(balance.unconfirmed, 1012345678);
+  });
+
+  it('should mine a block and confirm a tx', async () => {
+    const block = await miner.mineBlock();
+    assert(block);
+    await chain.add(block);
+
+    // Check SPV & Full nodes are in sync
+    await event(spvnode, 'block');
+    assert.deepStrictEqual(node.chain.tip, spvnode.chain.tip);
+
+    // Check SPV wallet balance
+    await event(spvwallet, 'balance');
+    const balance = await spvwallet.getBalance();
+    assert.strictEqual(balance.confirmed, 1012345678);
+  });
+
+  it('should handle a reorg', async () => {
+    assert.strictEqual(chain.height, 12);
+
+    // Main chain is ahead by 1 block now, catch the alt chain up
+    const entry = await chain.getEntry(tip2.hash);
+    const block1 = await miner.mineBlock(entry);
+    await chain.add(block1);
+    const entry1 = await chain.getEntry(block1.hash());
+    assert(entry1);
+
+    // Tie game!
+    assert.strictEqual(chain.height, entry1.height);
+
+    // Now reorg main chain by adding a block to alt chain
+    const block2 = await miner.mineBlock(entry1);
+    assert(block2);
+
+    let forked = false;
+    chain.once('reorganize', () => {
+      forked = true;
+    });
+
+    await chain.add(block2);
+
+    assert(forked);
+    assert.bufferEqual(chain.tip.hash, block2.hash());
+    assert(chain.tip.chainwork.gt(tip1.chainwork));
+
+    // Give SPV node a second to catch up before checking sync with fullnode.
+    // Waiting for specific events like 'block', 'full', or 'tip' is hard here
+    // because we don't really know when we are at the proper tip.
+    await new Promise(r => setTimeout(r, 5000));
```

Comment:
Maybe we can wait for an event with a specific value?

Reply:
Tried a few things with `while` loops waiting for a certain chain height, can't really figure out a more elegant way to do this... any thoughts?

Reply:
What about something like this?
```js
async function eventValue(obj, name, expected, timeout = 10000) {
  assert(typeof obj === 'object');
  assert(typeof name === 'string');

  return new Promise((resolve, reject) => {
    let timer = null;

    const check = (value) => {
      if (value === expected) {
        obj.removeListener(name, check);
        clearTimeout(timer);
        resolve();
      }
    };

    obj.addListener(name, check);

    timer = setTimeout(() => {
      obj.removeListener(name, check);
      reject(new Error('Timeout waiting for event value.'));
    }, timeout);
  });
}
```
I have been using this in https://github.com/bcoin-org/bcoin/pull/758 to wait for values:
```js
async function forValue(obj, key, val, timeout = 60000) {
  assert(typeof obj === 'object');
  assert(typeof key === 'string');

  const ms = 10;
  let interval = null;
  let count = 0;

  return new Promise((resolve, reject) => {
    interval = setInterval(() => {
      if (obj[key] === val) {
        clearInterval(interval);
        resolve();
      } else if (count * ms >= timeout) {
        clearInterval(interval);
        reject(new Error('Timeout waiting for value.'));
      }
      count += 1;
    }, ms);
  });
};
```